internal/compiler/js_diagnostics.go

```diff
@@ -0,0 +1,526 @@
+package compiler
+
+import (
+	"github.com/microsoft/typescript-go/internal/ast"
+	"github.com/microsoft/typescript-go/internal/binder"
+	"github.com/microsoft/typescript-go/internal/core"
+	"github.com/microsoft/typescript-go/internal/diagnostics"
+	"github.com/microsoft/typescript-go/internal/scanner"
+)
+
+// jsDiagnosticsVisitor is used to find TypeScript-only constructs in JavaScript files
+type jsDiagnosticsVisitor struct {
+	sourceFile  *ast.SourceFile
+	diagnostics []*ast.Diagnostic
+}
+
+// getJSSyntacticDiagnosticsForFile returns diagnostics for TypeScript-only constructs in JavaScript files
+func (p *Program) getJSSyntacticDiagnosticsForFile(sourceFile *ast.SourceFile) []*ast.Diagnostic {
+	if result, ok := p.jsDiagnosticCache.Load(sourceFile); ok {
+		return result
+	}
+
+	visitor := &jsDiagnosticsVisitor{
+		sourceFile:  sourceFile,
+		diagnostics: []*ast.Diagnostic{},
+	}
+
+	// Walk the entire AST to find TypeScript-only constructs
+	visitor.walkNodeForJSDiagnostics(sourceFile.AsNode(), sourceFile.AsNode())
+
+	diagnostics, _ := p.jsDiagnosticCache.LoadOrStore(sourceFile, visitor.diagnostics)
+	return diagnostics
```

Comment:
Extract the visitor and visitor call into a helper function `getJSSyntacticDiagnosticsForFile` and then change this to be:

```suggestion
	diagnostics, _ := p.jsDiagnosticCache.LoadOrStore(sourceFile, getJSSyntacticDiagnosticsForFile(sourceFile))
	return diagnostics
```